Marlin/Marlin_main.cpp

```diff
@@ -4464,6 +4467,128 @@ inline void gcode_G92() {
+  }
+  
+  inline void delay_for_spindle_power_down() {
+    refresh_cmd_timeout();                                                       // wait for spindle to stop turning
+    while (PENDING(millis(), SPINDLE_POWER_DOWN_DELAY * 1000 + previous_cmd_ms + 1)) idle(); 
+  }
+
+  inline void gcode_M3_M4() {
+
+    // this is the "pretty" version.  It accepts RPM as input
+    #if ENABLED(SPINDLE_SPEED)
+  
+      // use gcode_M4() to get the points needed to compute the RPM vs ocr_val line
+  
+      // Our final value for ocr_val is an unsigned 8 bit value between 0 and 255 which usually means uint8_t.
+      // Went to uint16_t because some of the uint8_t calculations would sometimes give 1000 0000 rather than 1111 1111.
+      // Then needed to AND the uint16_t result with 0xffff to make sure we only wrote the byte of interest.
+  
+      float spindle_speed = code_seen('S') ? code_value_float() : 0; 
+      if (spindle_speed == 0) { 
+        stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle off
+        WRITE(SPINDLE_ENABLE_PIN, !SPINDLE_ENABLE_INVERT);  //turn spindle off (active low);
+        delay_for_spindle_power_down();
+      }
+      else {
+        int16_t ocr_val = 2.55 * (spindle_speed - RPM_INTERCEPT) / RPM_SLOPE;      // convert RPM to PWM duty cycle
+        float temp = 2.55 * (spindle_speed - RPM_INTERCEPT) / RPM_SLOPE;
+        if (temp >= 255)                ocr_val = 255;                                        //limit to max the Atmel PWM will support
+        if (spindle_speed <= RPM_MIN)   ocr_val = 2.55 * (RPM_MIN - RPM_INTERCEPT) / RPM_SLOPE; // minimum setting
+        if (spindle_speed >= RPM_MAX)   ocr_val = 2.55 * (RPM_MAX - RPM_INTERCEPT) / RPM_SLOPE; // limit to max RPM
+        if (SPINDLE_SPEED_INVERT)       ocr_val = 255 - ocr_val; 
+        stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle on
+        WRITE(SPINDLE_ENABLE_PIN, SPINDLE_ENABLE_INVERT);  // turn spindle on (active low)
+        analogWrite(SPINDLE_SPEED_PIN, ocr_val & 0xFF);  //only write lowest byte 
+        delay_for_spindle_power_up();
+      } 
+    #else
+      stepper.synchronize();   // wait until previous movement commands (G0, G1, G2 & G3) have completed before turning spindle on
+      WRITE(SPINDLE_ENABLE_PIN, SPINDLE_ENABLE_INVERT);  //turn spindle on (active low) if spindle speed option not enabled
+      delay_for_spindle_power_up();
+    #endif
+  } 
+ 
+/**
+ * gcode_M4 is used for debugging and to get the points needed to compute the RPM vs ocr_val line
+ * 
+ * it accepts inputs of 0-255
+ *
+ * To use it you'll need to change the M4 command from "gcode_M3_M4()" to "gcode_M4()"
+ *   
+ * This was used to get the RPM vs. OCRxx PWM values
+ * That was thrown into Excel to determine a best fit line
+ * for use in the M3 routine.
+ */
+
+  inline void gcode_M4() {
```

Comment:
Don't name this `gcode_M4` because it's not a formal handler, nor require users to hack code to debug things. It's better if you make a settings flag to do this instead, and name the function something like `debug_G4`.